src/IJ_mv/IJMatrix_parcsr.c

```diff
@@ -2912,6 +2912,11 @@ hypre_IJMatrixAssembleParCSR(hypre_IJMatrix *matrix)
       hypre_IJMatrixAssembleFlag(matrix) = 1;
    }
 
+   /* Generate the nonzero rows in the diag and offd matrices */
```

Comment:
What's the reason of moving these two here?

Reply:
When we want to set the exact number of nonzeros of the diagonal and off-diagonal blocks of a `ParCSRMatrix`, which can be done via `HYPRE_IJMatrixSetDiagOffdSizes`, this function turns off the `NeedAux` flag, see: https://github.com/hypre-space/hypre/blob/f7787ab0aed34022503c656a3efc338e4616ff21/src/IJ_mv/IJMatrix_parcsr.c#L168

Fast-forward to `HYPRE_IJMatrixAssemble`, looking at the code on master, the following calls:
https://github.com/hypre-space/hypre/blob/f7787ab0aed34022503c656a3efc338e4616ff21/src/IJ_mv/IJMatrix_parcsr.c#L2865-L2866

are under the `if` statement https://github.com/hypre-space/hypre/blob/f7787ab0aed34022503c656a3efc338e4616ff21/src/IJ_mv/IJMatrix_parcsr.c#L2642

and thus are not reached. So, if want to build the `rownnz` info, we want them to be outside the `if` statement involving `NeedAux`. 

PS: It's actually a bit more correct to keep the `rownnz` functions inside the `if` block: https://github.com/hypre-space/hypre/blob/f7787ab0aed34022503c656a3efc338e4616ff21/src/IJ_mv/IJMatrix_parcsr.c#L2634

I implemented this minor fix on [352f74a](https://github.com/hypre-space/hypre/pull/599/commits/352f74a67a11d61ea473b74b990c8cd86d8527c9)

Reply:
Thanks! I was just curious. Looks good!